oidc-exchange.py

```diff
@@ -0,0 +1,137 @@
+
+
+repository_url = os.getenv("INPUT_REPOSITORY_URL")
+if not repository_url:
+    # Easy case: no explicit repository URL, which means we're using PyPI and we can just
+    # hardcode the exchange endpoint and OIDC audience.
+    token_exchange_url = "https://pypi.org/_/oidc/github/mint-token"
+    oidc_audience = "pypi"
+else:
+    repository_domain = urlparse(repository_url).netloc
+    token_exchange_url = f"https://{repository_domain}/_/oidc/github/mint-token"
+
+    # Indices are expected to support `https://{domain}/_/oidc/audience`,
+    # which tells OIDC exchange clients which audience to use.
+    audience_url = f"https://{repository_domain}/_/oidc/audience"
+    audience_resp = requests.get(audience_url)
+
+    if not audience_resp.ok:
+        if audience_resp.status_code == 403:
+            # This index supports OIDC, but forbids the client from using
+            # it (either because it's disabled, ratelimited, etc.)
+            die(
+                f"audience retrieval failed: repository at {repository_domain} has OIDC disabled"
+            )
+        elif audience_resp.status_code == 404:
+            # This index does not support OIDC.
+            die(
+                "audience retrieval failed: repository at "
+                f"{repository_domain} does not indicate OIDC support"
+            )
+        else:
+            # Unknown: the index may or may not support OIDC, but didn't respond with
+            # something we expect. This can happen if the index is broken, in maintenance mode,
+            # misconfigured, etc.
+            die(
+                "audience retrieval failed: repository at "
+                f"{repository_domain} responded with unexpected {audience_resp.status_code}"
+            )
+
+    oidc_audience = audience_resp.json()["audience"]
+
+debug(f"selected OIDC token exchange endpoint: {token_exchange_url}")
+
+try:
+    oidc_token = id.detect_credential(audience=oidc_audience)
+except id.IdentityError as exc:
+    die(
+        dedent(
+            f"""
+            OIDC token retrieval failed: {exc}
+
+            This generally indicates a workflow configuration error, such as insufficient
+            permissions. Make sure that your workflow has `id-token: write` configured
+            at either the workflow or job level, e.g.:
+
+            ```yaml
+            permissions:
+              id-token: write
+            ```
+            """
+        )
+    )
+
+# Now we can do the actual token exchange.
+mint_token_resp = requests.post(
+    token_exchange_url,
+    json={"token": oidc_token},
+)
+
+if not mint_token_resp.ok:
+    message = "Token request failed"
+    try:
+        error_payload = mint_token_resp.json()
+        reasons = "\n".join(
+            f"* `{error['code']}`: {error['description']}"
+            for error in error_payload["errors"]
+        )
+        message = f"""
+{message}: the server refused the request for the following reasons:
+
+{reasons}
```

Comment:
Reflowed, but I can't actually `dedent` here because of the `reasons` -- they're newline-delimited, so `dedent` won't remove the indentation correctly 😞